rebuild/src/public_key.h

```diff
@@ -0,0 +1,96 @@
+#ifndef BLST_TS_PUBLIC_KEY_H__
+#define BLST_TS_PUBLIC_KEY_H__
+
+#include <memory>
+
+#include "addon.h"
+#include "blst.hpp"
+#include "napi.h"
+
+#define BLST_TS_PUBLIC_KEY_LENGTH_COMPRESSED 48U
+#define BLST_TS_PUBLIC_KEY_LENGTH_UNCOMPRESSED 96U
+
+typedef struct {
+    std::unique_ptr<blst::P1_Affine> smart_pointer;
```

Comment:
Can we think of a more meaningful name? And in general, why do we need both a unique pointer and a raw pointer? I lack domain knowledge here, sadly.

Reply:
Yes.  Was the best way I could solve the issue.  In some instances just a copy to the existing raw pointer of a blst:point is needed.  In other code paths a blst::point needs to be created and the lifetime managed for during the function invocation. so on the paths were just a copy of the pointer (a point already managed by one of the wrapper classes that has the smart pointer) the smart pointer is not used.  On code paths where a wrapper is not available (a serialized key or signature was passed to the function) a smart pointer is used to manage the lifetime of the point and the `raw_pointer` points to the value held by the smart pointer.  Also helps to manage point type conversions.  Example is a P1->P1Affine.  We store the keys in P1 but verification requires an Affine version so a new point must be created and it gets thrown away at the end of the function invocation.  There are instances where the PublicKey that is getting passed in is already in Affine form so its a P1Affine->P1Affine and no conversion is necessary so in those cases just the raw pointer is passed in to verify and no new point need to be saved in the smart pointer.  fyi verify is a const function so no mutation.  

```c++
        Napi::Value sig_val = info[2];
        P2AffineGroup sig_point;
        if (sig_val.IsTypedArray()) {
            Napi::Uint8Array typed_array = sig_val.As<Napi::Uint8Array>();
            std::string err_out{"BLST_ERROR: SignatureArg"};
            if (!is_valid_length(********)) {
                Napi::TypeError::New(env, err_out).ThrowAsJavaScriptException();
                return env.Undefined();
            } else {
                sig_point.smart_pointer = std::make_unique<blst::P2_Affine>(
                    typed_array.Data(), typed_array.ByteLength());
                sig_point.raw_point = sig_point.smart_pointer.get();
            }
        } else {
            Signature *to_verify =
                Signature::Unwrap(sig_val.As<Napi::Object>());
            sig_point = to_verify->_point->AsAffine();
        }
        ```

Using this pattern avoids the share_ptr atomics delays.

I suppose it could be refactored though so that a shared_ptr is used instead of unique_ptr for all of the wrapper classes and then when a byte array (serialized key or signature) is passed to the function a shared_ptr (that will be unique) will be created....

Reply:
In my opinion, it's more practical to focus on correctness and simplicity rather than optimize away shared pointers. I'd argue that the cost of an atomic counter is negligible to the potential cost of a segfault (or worse) if someone touches the code without verifying everything.

Then, if performance is of an issue, profile it to find the real bottlenecks.

Reply:
After I get this work merged I will make a trial branch for this comment and [this comment](https://github.com/ChainSafe/blst-ts/pull/119#discussion_r1521824150) as they are related.  I want to perf test the shared_ptr but there are enough changes with all this stuff that it will not be clear what we are looking at.  TBD